src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -300,6 +304,158 @@ topSort g = runContT (evalStateT (topSort' g) initialState) id where
+-- scc (3 * 1 * 4 * 1 * 5) == 'edges'  [ (NonEmpty.'NonEmpty.vertex'  3      , NonEmpty.'NonEmpty.vertex'  5      )
+--                                   , (NonEmpty.'NonEmpty.vertex'  3      , NonEmpty.'NonEmpty.clique1' [1,4,1])
+--                                   , (NonEmpty.'NonEmpty.clique1' [1,4,1], NonEmpty.'NonEmpty.vertex'  5      ) ]
+-- 'isAcyclic' . scc == 'const' True
+-- 'isAcyclic' x     == (scc x == 'gmap' NonEmpty.'NonEmpty.vertex' x)
+-- @
+scc :: AdjacencyIntMap -> AM.AdjacencyMap AdjacencyIntMap
+scc g = condense g $ execState (gabowSCC g) initialState where
+  initialState = C 0 0 [] [] IntMap.empty IntMap.empty IntMap.empty [] []
+
+
+-- StateSCC is the data type used to keep track of state while
+-- calculating the strongly connected components of a graph with
+-- gabowSCC.
+type PreorderId = Int
+type ComponentId = Int
+
+data StateSCC
+  = C { -- on entry, a vertex is given a preorder id to be stored in
+        -- the preorders table.
+        preorder      :: !PreorderId
+        
+        -- when a scc is found, its vertices are assigned a component
+        -- id.
+      , component     :: !ComponentId
+      
+        -- the so called boundary stack of Gabow's algorithm. We store
+        -- (preorder id, vertex) pairs to avoid log(n) cost in lookup
+        -- when calling popBoundary.
+      , boundary      :: ![(PreorderId,Int)]
+      
+        -- the so called path stack of Gabow's algorithm. It is a list
+        -- of vertices in the path of the dfs search tree during
+        -- traversal. Vertices are popped when a new scc is found.
+      , dfsPath       :: ![Int]
+      
+        -- the table storing preorder ids for vertices.
+      , preorders     :: !(IntMap.IntMap PreorderId)
+      
+        -- the table storing the component id to which vertices have
+        -- been assigned.
+      , components    :: !(IntMap.IntMap ComponentId)
+      
+        -- the remaining fields are used for calculating the
+        -- condesation of the traversed graph.  inner_graphs stores a
+        -- table of strongly connected components indexed by their
+        -- component id.  inner_edges holds a list of edges that has
+        -- been classified as inner during the traversal. They are
+        -- stored with a preorderId of the target vertex to
+        -- distinguish which connected component each vertex belongs
+        -- to. outer_edges holds a list of edges classified as between
+        -- strongly connected components. It and inner_graphs are lazy
+        -- because they are not always required to compute the
+        -- condensation.
+      , inner_graphs  :: IntMap.IntMap (List AdjacencyIntMap)
+      , inner_edges   :: ![(PreorderId,(Int,Int))]
+      , outer_edges   :: [(Int,Int)]
+      } deriving (Show)
+
+gabowSCC :: AdjacencyIntMap -> State StateSCC ()
+gabowSCC g =
+  do let adjacent = IntSet.toList . flip postIntSet g
```

Comment:
Could you avoid the construction of an intermediate list here? Instead of `forM_` you could probably use `IntSet.foldr`.

(It may well be the case that the intermediate list is optimised away, but it still feels better to use the set API directly.)

Reply:
That sounds good, but `adjacent` is used like `forM_ (adjacent v) ...` so I think we'd need some `foldM`, which I don't think is exported by either set module? Might it be worth adding `{-# inline ... #-}` declarations?

Reply:
Do you mean you need something like `forEach`/`forEachInt`?

```haskell
forEach :: Applicative f => Set a -> (a -> f b) -> f ()
forEach set f = Set.foldr (\x y -> f x *> y) (pure ()) set

... forEach (postSet g u) $ \v -> ...
```

If yes, feel free to add these auxiliary functions into `Algebra.Graph.Internal`.

Reply:
(Or their strict versions if need be.)

Reply:
Sounds good! I added them as `setForEach_` and `intsetForEach_`. Maybe the under bars are not needed/too ugly?

Reply:
I would just go for `forEach` and `forEachInt` -- these names should read well in the code.